cpmpy/transformations/normalize.py

```diff
@@ -34,3 +37,118 @@ def unravel(lst, append):
+                cond, bool_expr = args
+                if is_false_cst(cond) or is_true_cst(bool_expr):
+                    newlist.append(BoolVal(True))
+                elif is_true_cst(cond):
+                    newlist.append(bool_expr)
+                elif is_false_cst(bool_expr):
+                    newlist += simplify_boolean([~cond])
+                else:
+                    newlist.append(cond.implies(bool_expr))
+
+            elif expr.name == "not":
+                if is_true_cst(args[0]):
+                    newlist.append(BoolVal(False))
+                elif is_false_cst(args[0]):
+                    newlist.append(BoolVal(True))
+                else:
+                    newlist.append(~args[0])
+
+            else: # numerical expressions
+                newlist.append(Operator(expr.name, args))
+
+        elif isinstance(expr, Comparison):
+            lhs, rhs = simplify_boolean(expr.args, num_context=True)
+            name = expr.name
+            if isinstance(lhs, int) and is_boolexpr(rhs): # flip arguments of comparison to reduct nb of cases
+                flipmap = {"==":"==", "!=":"!=", "<=":">=", "<":">"}
+                name = flipmap[name]
+                lhs, rhs = rhs, lhs
+            """
+            Simplify expressions according to this table:
+            x  | <0	 0	 1	>1
+            --------------------
+            == |  F	~x	 x	 F
+            != |  T	 x	~x	 T
+            >  |  T	 x	 F	 F
+            <  |  F	 F	~x	 T
+            >= |  T	 T	 x	 F
+            <= |  F	~x	 T	 T
+            """
+            if is_boolexpr(lhs) and isinstance(rhs, int):
+                # direct simplification of boolean comparisons
+                if rhs < 0:
+                    newlist.append(BoolVal(name in  {"!=", ">", ">="})) # all other operators evaluate to False
+                if rhs == 0:
+                    if name == "!=" or name == ">":
+                        newlist.append(lhs)
+                    if name == "==" or name == "<=":
+                        newlist.append(~lhs)
+                    if name == "<":
+                        newlist.append(BoolVal(False))
+                    if name == ">=":
+                        newlist.append(BoolVal(True))
+                if rhs == 1:
+                    if name == "==" or name == ">=":
+                        newlist.append(lhs)
+                    if name == "!=" or name == "<":
+                        newlist.append(~lhs)
+                    if name == ">":
+                        newlist.append(BoolVal(False))
+                    if name == "<=":
+                        newlist.append(BoolVal(True))
+                if rhs > 1:
+                    newlist.append(BoolVal(name in  {"!=", "<", "<="})) # all other operators evaluate to False
+            else:
+                newlist.append(eval_comparison(name, lhs, rhs))
+        elif hasattr(expr, "decompose"):
+            expr = copy.deepcopy(expr)
```

Comment:
a deepcopy will copy the args and all children of the args.

But here, we 'recreate' the args ourselves, so a shallow copy where we overwrite the args is what we need (otherwise could be useless performance penalty)